sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -230,6 +236,7 @@ case class HashAggregateExec(
          | private void $doAgg() throws java.io.IOException {
          |   // initialize aggregation buffer
          |   $initBufVar
+         |   $hasInput = false;
```

Comment:
How about using a returned value of `doAgg` to check `hasInput`? I feel it's better to avoid mutable states as much as possible.

Reply:
I actually tried it, but found it was hard because we don't control the codegen of line `/* 038 */       if (shouldStop()) return;` in `doAgg`.

Reply:
btw, if `resultExpressions` is empty, we need  to check `hasInput`? It seems it is ok to return an empty iterator in that case?

Reply:
If the result expression is empty and `hasInput` is `false`, it should return an empty iterator (this is exactly the corner case we want to fix). If the result expression is empty but `hasInput` is `true`, it should append empty rows.